org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/internal/util/progress/ProblemsReportingProgressListener.java

```diff
@@ -48,69 +48,117 @@ public ProblemsReportingProgressListener(InternalGradleBuild gradleBuild) {
     public void statusChanged(ProgressEvent event) {
         if (event instanceof ProblemEvent) {
             ProblemEvent problemEvent = (ProblemEvent) event;
-            BaseProblemDescriptor eventDescriptor = problemEvent.getDescriptor();
             try {
-                if (eventDescriptor instanceof ProblemDescriptor) {
-                    reportProblem((ProblemDescriptor) eventDescriptor);
-                } else if (eventDescriptor instanceof ProblemAggregationDescriptor) {
-                    for (ProblemAggregation aggregation : ((ProblemAggregationDescriptor) eventDescriptor).getAggregations()) {
-                        for (ProblemDescriptor descriptor : aggregation.getProblemDescriptors()) {
-                            reportProblem(descriptor);
-                        }
-                    }
+                if (problemEvent instanceof SingleProblemEvent) {
+                    reportProblem((SingleProblemEvent) problemEvent);
+                } else if (problemEvent instanceof ProblemAggregationEvent) {
                 }
             } catch (Exception e) {
                 CorePlugin.logger().warn("Cannot report problem " + problemEvent, e);
             }
         }
     }
 
-    private void reportProblem(ProblemDescriptor descriptor) {
-        Optional<Pair<IResource,Integer>> location = resourceAndFileNumberOfFirstFileLocation(descriptor);
-        if (location.isPresent()) {
-            GradleErrorMarker.createMarker(
-                toMarkerSeverity(descriptor.getSeverity()),
-                location.get().getFirst(), this.gradleBuild,
-                descriptor.getLabel().getLabel(),
-                null, // TODO (donat) Gradle 8.7 descriptor.getException().getException(),
-                location.get().getSecond(),
-                toPath(descriptor.getCategory()),
-                descriptor.getSolutions().stream().map(Solution::getSolution).collect(Collectors.toList()),
-                descriptor.getDocumentationLink().getUrl()
-            );
-        } else {
-            GradleErrorMarker.createMarker(
-                toMarkerSeverity(descriptor.getSeverity()),
-                ResourcesPlugin.getWorkspace().getRoot(),
-                this.gradleBuild,
-                descriptor.getLabel().getLabel(),
-                null, // TODO (donat) Gradle 8.7 descriptor.getException().getException(),
-                -1,
-                toPath(descriptor.getCategory()),
-                descriptor.getSolutions().stream().map(Solution::getSolution).collect(Collectors.toList()),
-                descriptor.getDocumentationLink().getUrl()
-            );
+    private void reportProblem(SingleProblemEvent event) {
+        List<Location> locations = event.getLocations();
+
+        // 1/4 offset in file location
+        Optional<OffsetInFileLocation> offsetInFileLocation = locations.stream().filter(OffsetInFileLocation.class::isInstance).map(OffsetInFileLocation.class::cast).findFirst();
```

Comment:
🤔 maybe put this in a method, including the `.isPresent()`part?

`hasLocation(OffsetInFileLocation.class)`

Reply:
I've refactored this part.